src/data_model.rs

```diff
@@ -0,0 +1,105 @@
+/// The size of a type.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[allow(missing_docs)]
+pub struct Size {
+    /// The size in bits of this type.
+    bits: u8,
+    /// The number of bits in a byte.
+    char_bit: u8,
+}
+
+impl Size {
+    /// Return the number of bits this `Size` represents.
+    pub fn bits(self) -> u8 {
+        self.bits
+    }
+
+    /// Return the number of bytes in a size.
+    ///
+    /// Depending on the target, a byte may not necessarily be 8 bits.
+    pub fn bytes(self) -> u8 {
+        let bytes = self.bits / self.char_bit;
+        // make sure that self.bits is a multiple of char_bit
+        assert_eq!(bytes * self.char_bit, self.bits);
+        bytes
+    }
+    /// Create a new `Size`, assuming that `char_bit` is 8.
+    fn with_bits(bits: u8) -> Self {
```

Comment:
If we're going to support `CHAR_BIT!=8`, it seems like we should consistently avoid assuming `CHAR_BIT=8`. How inconvenient would it be to add a `char_bit` parameter here?

Reply:
My idea was that this would be an internal API and that CHAR_BIT would default to 8. You can still construct `Size` with a different bit width manually.

However, after looking at your comments below I'm starting to wonder the utility of having a configurable CHAR_BIT. If we can't even name `bytes()` after the number of bits in a byte, is it worth having the distinction? Do you know any targets rustc supports with CHAR_BIT != 8?

Reply:
No, I don't know of any targets in rustc with `CHAR_BIT != 8`. I think it is reasonable to have the API assume `CHAR_BIT == 8`, but I'd like to either make that assumption everywhere, or not at all. If we assume `CHAR_BIT` is 8, then `Size` shouldn't have a `char_bit` field. If we don't assume it, then we shouldn't have any implicit 8-bit assumptions.

Reply:
I'm ok with getting rid of char_bit, but in that case we should either change Size to `enum { 8, 16, 32, 64 }` or return an `Option` from bytes. If there's a size that's not a multiple of 8, there's nothing to do right now except panic.

Reply:
Actually I guess there is a 3rd option which is to round up to the nearest multiple of 8.

Reply:
A panic sounds reasonable to me here. An enum would also be fine.

Reply:
Ok, I changed it to an enum.